File: 108.convert-sorted-array-to-binary-search-tree.py
#
# @lc app=leetcode id=108 lang=python3
#
# [108] Convert Sorted Array to Binary Search Tree
#
from typing import List
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# @lc code=start
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        return self.repeat(nums)

    def repeat(self, nums: List[int]) -> TreeNode:
        if len(nums) == 0:
            return None
        half = len(nums) // 2

        left = nums[:half]
        center = nums[half]
        right = nums[half + 1:]

        node = TreeNode(center)
        node.left = self.repeat(left)
        node.right = self.repeat(right)

        return node

# @lc code=end

s = Solution()
s.sortedArrayToBST([-10,-3,0,5,9])

File: 108.convert-sorted-array-to-binary-search-tree.js
/*
 * @lc app=leetcode id=108 lang=javascript
 *
 * [108] Convert Sorted Array to Binary Search Tree
 */

// Definition for a binary tree node.
class TreeNode {
  constructor(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
  }
}

// @lc code=start
/**
 * @param {array} arr
 * @return {[array, number, array]}
 */
const sliceArray = (arr) => {
  const centerIdx = Math.floor(arr.length / 2);
  
  const left = arr.slice(0, centerIdx);
  const center = arr[centerIdx];
  const right = arr.slice(centerIdx + 1, arr.length);

  return [left, center, right];
};

/**
 * @param {number[]} nums
 */
const makeBST = (nums) => {
  if (nums.length) {
    const [left, center, right] = sliceArray(nums);
    const node = new TreeNode(center, makeBST(left), makeBST(right));

    return node;
  }
};

/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
const sortedArrayToBST = (nums) => {
  return makeBST(nums);
};
// @lc code=end

